nsfc-plugin/src/main/java/org/osc/controller/nsfc/api/NeutronSfcSdnRedirectionApi.java

```diff
@@ -69,15 +77,73 @@ public InspectionPortElement getInspectionPort(InspectionPortElement inspectionP
             return null;
         }
 
-        return null; // TODO (Dmitry) Implement
+        String portPairId = inspectionPort.getElementId();
+        PortPair portPair = null;
+
+        if (portPairId != null) {
+            portPair = this.osCalls.getPortPair(portPairId);
+        }
+
+        if (portPair == null) {
+            LOG.warn("Failed to retrieve InspectionPort by id! Trying by ingress and egress " + inspectionPort);
+
+            NetworkElement ingress = inspectionPort.getIngressPort();
+            NetworkElement egress = inspectionPort.getEgressPort();
+
+            portPair = this.utils.fetchInspectionPortByNetworkElements(ingress, egress);
+        }
+
+        if (portPair != null) {
+            // only id is ever used
+            return new InspectionPortEntity(portPairId, null, null, null);
+        }
+
+        return null;
     }
 
     @Override
     public Element registerInspectionPort(InspectionPortElement inspectionPort) throws Exception {
         if (inspectionPort == null) {
             throw new IllegalArgumentException("Attempt to register null InspectionPort");
         }
-        return null; // TODO (Dmitry) Implement
+        PortPairGroup portPairGroup = null;
+        String inspectionPortGroupId = inspectionPort.getParentId();
+
+        if (inspectionPortGroupId != null) {
+            portPairGroup = this.osCalls.getPortPairGroup(inspectionPortGroupId);
+            checkArgument(portPairGroup != null && portPairGroup.getId() != null,
+                    "Cannot find %s by id: %s!", "Port Pair Group", inspectionPortGroupId);
+        }
+
+        NetworkElement ingress = inspectionPort.getIngressPort();
+        NetworkElement egress = inspectionPort.getEgressPort();
+        PortPair portPair = this.utils.fetchInspectionPortByNetworkElements(ingress, egress);
```

Comment:
do we need this check? cant we directly create portpair with the ingress and egress port?

SFC already validates if a port pair exists but not sure if it the call is idempotent.
https://github.com/openstack/networking-sfc/blob/659b5ee5349157aa6f313d866e527ef652219ace/networking_sfc/db/sfc_db.py#L509

check other places for ppg etc as well.

Reply:
That code raises an exception if the port pair protecting the given ingress/egress already exists. So, this code (which is a translation of the [old code with h2](https://github.com/opensecuritycontroller/sdn-controller-nsfc-plugin/blob/7eb78dc944ad239b54848943b15506c00fee6c5c/src/main/java/org/osc/controller/nsfc/api/NeutronSfcSdnRedirectionApi.java#L107) ) only creates a port pair if it doesn't yet exist.

Reply:
i understand. If we can rely on the SFC code to make this validation instead of relying on our own validation its less code to write and test.

Again it would only work if this specific call was idempotent. Can you try it out?